components/heap/heap_trace_standalone.c

```diff
@@ -304,46 +345,187 @@ static IRAM_ATTR void record_free(void *p, void **callers)
+            rCur->prev = rCur - 1;
+        }
+    }
+}
+
+/* remove the record from the linked list */
+static IRAM_ATTR void linked_list_remove(records_t* rs, heap_trace_record_t* rRemove)
 {
-    if (index < r->count - 1) {
-        // Remove the buffer entry from the list
-        memmove(&r->buffer[index], &r->buffer[index+1],
-                sizeof(heap_trace_record_t) * (r->capacity - index - 1));
+    // set as available
+    rRemove->address = 0;
+    rRemove->size = 0;
+
+    if (rs->count > 1) {
+
+        heap_trace_record_t* rPrev = rRemove->prev;
+        heap_trace_record_t* rNext = rRemove->next;
+
+        if (rPrev){
+            assert(rPrev->next == rRemove);
+        }
+
+        if (rNext){
+            assert(rNext->prev == rRemove);
+        }
+
+        // update prev neighbor
+        if (rPrev) {
+            rPrev->next = rNext;
+        }
+
+        // update next neighbor
+        if (rNext) {
+            rNext->prev = rPrev;
+        }
+
+        // update first
+        if (rRemove == rs->first) {
+            rs->first = rNext;
+        }
+
+        // update last
+        if (rRemove == rs->last) {
+            rs->last = rPrev;
+        }
+
+        // move after last
+        linked_list_move(rRemove, rs->last);
+
+        rs->count--;
+
+        if (rs->count == 1){
+            assert(rs->last == rs->first);
+        }
+
+    } else if (rs->count == 1) {
+        rs->count--;
+    }
+}
+
+// insert a record after the given record
+static IRAM_ATTR void linked_list_move(heap_trace_record_t* rInsert, heap_trace_record_t* rAfter)
+{
+    rAfter->next->prev = rInsert;
+    rInsert->prev = rAfter;
+    rInsert->next = rAfter->next;
+    rAfter->next = rInsert;
+}
+
+// get next available record
+static IRAM_ATTR heap_trace_record_t* linked_list_next_available(const records_t* rs)
+{
+    if (rs->count == rs->capacity){
+        return NULL;
+    }
+    if (rs->count == 0){
+        assert(rs->last == rs->first);
+        return rs->last;
+    } else {
+        assert(rs->last->next->address == NULL);
+        return rs->last->next;
+    }
+}
+
+// copy a record.
+// Note: only copies the *allocation data*, not the next & prev ptrs
+static IRAM_ATTR void linked_list_copy(heap_trace_record_t *rDest, const heap_trace_record_t *rSrc)
+{
+    rDest->ccount = rSrc->ccount;
+    rDest->address = rSrc->address;
+    rDest->size = rSrc->size;
+    memcpy(rDest->freed_by, rSrc->freed_by, sizeof(void *) * STACK_DEPTH);
+    memcpy(rDest->alloced_by, rSrc->alloced_by, sizeof(void *) * STACK_DEPTH);
+}
+
+// Append a record to the end of the linked list.
+// This deep copies rAppend into the linked list. 
+static IRAM_ATTR bool linked_list_append_copy(records_t* rs, const heap_trace_record_t *rAppend)
+{
+    if (rs->count < rs->capacity) {
+
+        // copy record into the next available slot
+        heap_trace_record_t* rAvailable = linked_list_next_available(r);
+        linked_list_copy(rAvailable, rAppend);
+
```

Comment:
I should add a comment here why we don't need to update the linked list connectivity.

reason: the linked list is already fully connected. It's just that some of its nodes are "empty". So we just copy the data into the last node.